docs/design/proposal-troubleshoot-consolidated-cli.md

```diff
@@ -0,0 +1,199 @@
+
+## Non-Goals
+
+## Background
+
+## High Level Design
+
+Functions of `support-bundle`, `preflight`, `analyze`, `redact`, and `sbctl` binaries/tools should be rolled together into a single `troubleshoot` CLI plugin that can perform all necessary functions of the same.
+
+`troubleshoot` CLI plugin should be able to report on the version that is installed in the CLI, and any support bundle generated with `troubleshoot` should report its build/version inside the archive it generates.
+
+## Detailed Design
+
+- TODO: specify the public APIs we expect users of Troubleshoot to consume, e.g.
+  - collect()
+  - redact()
+  - analyze()
+
+In the interest of being able to work on this quickly without breaking existing use-cases, a new `troubleshoot` command should be created. Utilizing cobra and viper best practices from the cobra.dev docs.
+
+sbctl should be migrated to the troubleshoot repository in a "lift and shift" operation to start with, allowing for it's methods and functions to be called by the new `troubleshoot` command
+
+### Public APIs
+
+The existing package entrypoint we use from the support-bundle command is supportbundle.CollectSupportBundleFromSpec()
+which is an end-to-end function that runs collectors, redactors and then analysis.
+
+The Consolidated cli should be responsible for handling the end-to-end running of this process in the `cmd` leaving the importable `pkg` free to focus on defining the individual collect, redact analyze steps.
+
+this allows us to move functions out from behind `CollectSupportBundleFromSpec` so they can be easier called from other projects importing troubleshoot.
+
+those functions could be then called as:
+
+- `troubleshoot.Collect()`
+- `troubleshoot.Redact()`
+- `troubleshoot.Analyze()`
+
+### Usage patterns
+
+- generate a support bundle
+
+  `troubleshoot collect supportbundle.yaml`
+
+  `troubleshoot collect supportbundle.yaml secrets/default/kotsadm-appslug-supportbundle`
+
+  `troubleshoot collect https://kots.io`
+
+- use a spec to return a go/no-go preflight outcome
+
+  `troubleshoot preflight spec.yaml`
+
+  This should not only clearly state any reasons for failing, but also use standardised exit codes that can be used by automation tools.
+
+- Interact with an existing support bundle with kubectl.
+
+  `troubleshoot inspect support-bundle-12-12-2001.tar.gz`
+
+  This should set the `KUBECONFIG` environment variable, as well as a `TROUBLESHOOT_SHELL` environment variable before spawning a subshell.
+  This is to allow prompts and shell environments to be able to detect that they're running in a troubleshoot spawned shell, much like tmux et-al.
+
```

Comment:
do we also want to include a 'serve' type of option for this outside the shell?  Not sure there's big value in it, we should maybe just document that folks can use the same KUBECONFIG env.

Reply:
hmm. yeh I can't see much value for use by a human but it could be useful for writing automation against.